sdk/dotnet/Pulumi/Resources/ComponentResourceOptions.cs

```diff
@@ -48,37 +50,48 @@ internal override ResourceOptions Clone()
         /// <item><description>
         /// <see langword="null"/> values in <paramref name="options2"/> will be ignored.
         /// </description></item>
+        /// <item><description>
+        /// Providers is a special case. Only one value per package will be in the resulting list.
+        /// Priority is given to values in <paramref name="options2"/>. If multiple providers for
+        /// the same package are present, later providers take priority.
+        /// </description></item>
         /// </list>
         /// </summary>
         public static ComponentResourceOptions Merge(ComponentResourceOptions? options1, ComponentResourceOptions? options2)
         {
             options1 = options1 != null ? CreateComponentResourceOptionsCopy(options1) : new ComponentResourceOptions();
             options2 = options2 != null ? CreateComponentResourceOptionsCopy(options2) : new ComponentResourceOptions();
 
-            ExpandProviders(options1);
-            ExpandProviders(options2);
-
             // first, merge all the normal option values over
             MergeNormalOptions(options1, options2);
 
-            options1.Providers.AddRange(options2.Providers);
-
-            if (options1.Providers.Count == 1)
-            {
-                options1.Provider = options1.Providers[0];
-                options1.Providers.Clear();
-            }
+            options1.Providers = MergeProviders(options1.Providers, options2.Providers);
 
             return options1;
-            
-            static void ExpandProviders(ComponentResourceOptions options)
-            {
-                if (options.Provider != null)
+        }
+
+        private static List<ProviderResource> MergeProviders(List<ProviderResource> prov1, List<ProviderResource> prov2 )
+        {
+            var l = new List<ProviderResource>();
+            var taken = new HashSet<string>();
+            Action<List<ProviderResource>> add = (list) => {
+                for(var i = list.Count-1; i >= 0; i--)
```

Comment:
Just a style question since I don't typically write dotnet, but why are you going through it in reverse and not using a foreach loop?

Reply:
I'm not using a `foreach` loop because I want to iterate through the list in reverse. Every method (as in OOP) I found for backwards iteration involved creating a stack. I saw no reason to pay for that. 

If you're asking why I'm iterating in reverse, it's because I want to give priority to later elements in the input lists. This makes it simpler with the `taken` filter.